File: web/src/services/api/timelogs.ts
import { http } from "./client";
import { ENDPOINTS } from "./endpoints";
import type { PersonSummary } from "./types/common";
import type { UUID } from "@/types/primitive";
import { DataCleaner } from "@/utils/protocol";
import type { ListResponse } from "@/types/pagination";
import type { NoteSummary } from "./notes";

// Types local to timelogs
interface AreaSummary {
  id: UUID;
  name: string;
  color?: string | null;
}

interface VisionSummary {
  id: UUID;
  name: string;
  status?: string | null;
  area_id?: UUID | null;
}

export interface TimelogTaskSummary {
  id: UUID;
  content: string;
  vision_id: UUID | null;
  status?: string;
  vision_summary?: VisionSummary | null;
}

export interface Timelog {
  id: UUID;
  title: string;
  start_time: string;
  end_time: string;
  area_id: UUID | null;
  task_id?: UUID | null;
  area_summary?: AreaSummary | null;
  tracking_method: string;
  location?: string | null;
  energy_level?: number | null;
  notes?: string | null;
  tags?: string[] | null;
  extra_data?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  persons?: PersonSummary[];
  // Preferred single associated task summary
  task?: TimelogTaskSummary | null;
  linked_notes?: NoteSummary[];
  linked_notes_count?: number;
}

export interface TimelogCreate {
  title: string;
  start_time: string;
  end_time: string;
  area_id: UUID | null;
  tracking_method?: string;
  location?: string;
  energy_level?: number;
  notes?: string;
  tags?: string[];
  extra_data?: Record<string, unknown>;
  task_id?: UUID | null; // single associated task
  person_ids?: UUID[];
  [key: string]: unknown;
}

export type TimelogUpdate = Partial<TimelogCreate>;

interface EnergyInjectionResult {
  vision_id: UUID;
  experience_gained: number;
  stage_evolved: boolean;
  new_stage: number;
  total_experience: number;
}

export interface TimelogWithEnergyResponse extends Timelog {
  energy_injections?: EnergyInjectionResult[];
}

export interface TimelogAdvancedSearchRequest {
  start_date: string;
  end_date?: string;
  area_id?: UUID | null;
  without_area?: boolean;
  area_name?: string | null;
  description_keyword?: string | null;
  task_id?: UUID | null;
  without_task?: boolean;
}

export interface TimelogAdvancedSearchMetadata {
  start_date?: string | null;
  end_date?: string | null;
  window_start?: string | null;
  window_end?: string | null;
  tracking_method?: string | null;
  area_id?: UUID | null;
  without_area?: boolean | null;
  area_name?: string | null;
  description_keyword?: string | null;
  task_id?: UUID | null;
  without_task?: boolean | null;
  limit?: number | null;
  returned_count?: number | null;
  total_count?: number | null;
  truncated?: boolean | null;
}

export type TimelogListResponse = ListResponse<
  Timelog,
  TimelogAdvancedSearchMetadata
>;

export type TimelogAdvancedSearchResponse = TimelogListResponse;

const TIMELOG_PAGE_SIZE = 500;
const MAX_TIMELOG_RANGE_PAGES = 100;

function toTimelogPayload(
  payload: TimelogCreate | TimelogUpdate,
): Record<string, unknown> {
  return {
    title: payload.title,
    start_time: payload.start_time,
    end_time: payload.end_time,
    tracking_method: payload.tracking_method ?? "manual",
    location: payload.location,
    energy_level: payload.energy_level,
    notes: payload.notes,
    area_id: payload.area_id,
    task_id: payload.task_id,
    person_ids: payload.person_ids,
  };
}

export const timelogsApi = {
  fetchRange: async (start: string, end: string, trackingMethod?: string) => {
    const items: Timelog[] = [];
    let firstResponse: TimelogListResponse | null = null;
    let page = 1;
    let totalCount = 0;
    let totalPages = 0;

    while (page <= MAX_TIMELOG_RANGE_PAGES) {
      const response = await http.get<TimelogListResponse>(
        ENDPOINTS.TIMELOGS.BASE,
        {
          window_start: start,
          window_end: end,
          tracking_method: trackingMethod,
          page,
          size: TIMELOG_PAGE_SIZE,
        },
      );

      firstResponse ??= response;
      items.push(...response.items);
      totalPages = response.pagination?.pages ?? 0;
      totalCount = response.pagination?.total ?? items.length;
      const backendTruncated = response.meta?.truncated === true;

      if (items.length >= totalCount) break;
      if (totalPages > 0 && page >= totalPages) break;
      if (!backendTruncated && response.items.length < TIMELOG_PAGE_SIZE) break;

      page += 1;
    }

    if (!firstResponse) {
      return {
        items: [],
        pagination: { page: 1, size: 0, total: 0, pages: 0 },
        meta: {
          returned_count: 0,
          total_count: 0,
          truncated: false,
        },
      };
    }

    const truncated = items.length < totalCount;
    return {
      items,
      pagination: {
        page: 1,
        size: items.length,
        total: totalCount,
        pages: truncated ? Math.max(totalPages, 1) : 1,
      },
      meta: {
        ...firstResponse.meta,
        returned_count: items.length,
        total_count: totalCount,
        truncated,
      },
    };
  },

  create: (payload: TimelogCreate) => {
    const cleanedData = DataCleaner.create(toTimelogPayload(payload));
    return http.post<TimelogWithEnergyResponse>(
      ENDPOINTS.TIMELOGS.BASE,
      cleanedData,
    );
  },

  batchCreate: (timelogs: TimelogCreate[]) => {
    const cleanedTimelogs = timelogs.map((timelog) =>
      DataCleaner.create(toTimelogPayload(timelog)),
    );

    return Promise.all(
      cleanedTimelogs.map((timelog) =>
        http.post<Timelog>(ENDPOINTS.TIMELOGS.BASE, timelog),
      ),
    ).then((createdTimelogs) => ({
      created_count: createdTimelogs.length,
      failed_count: 0,
      created_timelogs: createdTimelogs,
      errors: [],
    }));
  },

  update: (id: UUID, payload: TimelogUpdate) => {
    const cleanedData = DataCleaner.update(toTimelogPayload(payload));

    return http.patch<Timelog>(
      ENDPOINTS.TIMELOGS.BY_ID(id),
      cleanedData,
    );
  },

  quickEnd: (id: UUID) =>
    http.patch<Timelog>(ENDPOINTS.TIMELOGS.BY_ID(id), {
      end_time: new Date().toISOString(),
    }),

  delete: (id: UUID) => http.delete<void>(ENDPOINTS.TIMELOGS.BY_ID(id)),

  batchDelete: (eventIds: UUID[]) =>
    Promise.allSettled(
      eventIds.map((eventId) => http.delete<void>(ENDPOINTS.TIMELOGS.BY_ID(eventId))),
    ).then((results) => {
      const failedIds = eventIds.filter((_, index) => results[index].status === "rejected");
      return {
        deleted_count: eventIds.length - failedIds.length,
        failed_ids: failedIds,
        errors: [],
      };
    }),

  restore: (_id: UUID) =>
    Promise.reject(new Error("Restore is not supported by LifeOS timelogs yet.")),

  batchRestore: (eventIds: UUID[]) =>
    Promise.resolve({
      deleted_count: 0,
      failed_ids: eventIds,
      errors: ["Restore is not supported by LifeOS timelogs yet."],
    }),

  advancedSearch: async (params: TimelogAdvancedSearchRequest) => {
    const withoutArea =
      params.without_area ?? params.area_id === null;
    const withoutTask =
      params.without_task ?? params.task_id === null;
    const response = await http.get<TimelogAdvancedSearchResponse>(
      ENDPOINTS.TIMELOGS.BASE,
      {
        window_start: params.start_date,
        window_end: params.end_date,
        query: params.description_keyword ?? undefined,
        area_id: withoutArea
          ? undefined
          : (params.area_id ?? undefined),
        without_area: withoutArea || undefined,
        area_name: params.area_name ?? undefined,
        task_id: withoutTask ? undefined : (params.task_id ?? undefined),
        without_task: withoutTask || undefined,
        size: 500,
      },
    );
    const returnedCount = response.items.length;
    const totalCount = response.pagination?.total ?? returnedCount;
    return {
      ...response,
      meta: {
        ...response.meta,
        start_date: response.meta?.start_date ?? params.start_date,
        end_date:
          response.meta?.end_date ??
          (params.end_date ? params.end_date : null),
        window_start: response.meta?.window_start ?? params.start_date,
        window_end:
          response.meta?.window_end ?? (params.end_date ? params.end_date : null),
        area_id:
          response.meta?.area_id ?? params.area_id ?? null,
        without_area:
          response.meta?.without_area ?? withoutArea,
        area_name:
          response.meta?.area_name ?? params.area_name ?? null,
        description_keyword:
          response.meta?.description_keyword ??
          params.description_keyword ??
          null,
        task_id: response.meta?.task_id ?? params.task_id ?? null,
        limit: response.meta?.limit ?? 500,
        returned_count: response.meta?.returned_count ?? returnedCount,
        total_count: response.meta?.total_count ?? totalCount,
        truncated: response.meta?.truncated ?? returnedCount < totalCount,
      },
    };
  },

  batchUpdate: (params: {
    timelog_ids: UUID[];
    update_type: "persons" | "title" | "task" | "area";
    persons?: {
      mode: "add" | "replace" | "clear";
      person_ids: UUID[];
    };
    title?: {
      mode: "replace" | "find_replace";
      value: string;
      find?: string;
    };
    task?: {
      mode: "replace" | "clear";
      task_id?: UUID;
    };
    area?: {
      area_id: UUID | null;
    };
  }) =>
    http.post<{
      updated_count: number;
      unchanged_ids?: UUID[];
      failed_ids: UUID[];
      errors: string[];
    }>(ENDPOINTS.TIMELOGS.BATCH_UPDATE, params),
};

File: src/lifeos_web/routers/timelogs.py
"""Timelog endpoints for the local Web UI."""

from __future__ import annotations

import math
from datetime import date, datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from lifeos_cli.application.time_preferences import to_storage_timezone
from lifeos_cli.db.models.person_association import person_associations
from lifeos_cli.db.services import timelogs as timelog_services
from lifeos_cli.db.services.timelog_support import (
    TimelogBatchUpdateInput,
    TimelogCreateInput,
    TimelogListInput,
    TimelogQueryFilters,
    TimelogUpdateInput,
)
from lifeos_web.deps import get_db_session
from lifeos_web.schemas import (
    ListResponse,
    Pagination,
    TimelogBatchUpdate,
    TimelogCreate,
    TimelogUpdate,
)
from lifeos_web.serialization import to_jsonable

router = APIRouter(prefix="/timelogs", tags=["timelogs"])
SessionDep = Annotated[AsyncSession, Depends(get_db_session)]


def _timelog_payload(timelog: object) -> dict[str, object]:
    payload = to_jsonable(timelog)
    if not isinstance(payload, dict):
        raise TypeError("Timelog serialization did not produce a dictionary.")
    area_id = payload.get("area_id")
    if area_id is not None:
        payload["area_summary"] = {
            "id": area_id,
            "name": None,
            "color": None,
        }
    return payload


async def _batch_add_timelog_people(
    session: AsyncSession,
    *,
    timelog_ids: list[UUID],
    person_ids: list[UUID],
) -> dict[str, object]:
    unique_timelog_ids = list(dict.fromkeys(timelog_ids))
    rows = await session.execute(
        select(
            person_associations.c.entity_id,
            person_associations.c.person_id,
        ).where(
            person_associations.c.entity_type == "timelog",
            person_associations.c.entity_id.in_(unique_timelog_ids),
        )
    )
    people_by_timelog: dict[UUID, list[UUID]] = {
        timelog_id: [] for timelog_id in unique_timelog_ids
    }
    for timelog_id, person_id in rows.all():
        people_by_timelog.setdefault(timelog_id, []).append(person_id)

    updated_count = 0
    unchanged_ids: list[UUID] = []
    failed_ids: list[UUID] = []
    errors: list[str] = []
    for timelog_id in unique_timelog_ids:
        merged_person_ids = list(
            dict.fromkeys([*people_by_timelog.get(timelog_id, []), *person_ids])
        )
        if merged_person_ids == people_by_timelog.get(timelog_id, []):
            unchanged_ids.append(timelog_id)
            continue
        try:
            await timelog_services.update_timelog(
                session,
                timelog_id=timelog_id,
                changes=TimelogUpdateInput(person_ids=merged_person_ids),
            )
            updated_count += 1
        except (LookupError, ValueError) as exc:
            failed_ids.append(timelog_id)
            errors.append(str(exc))

    return {
        "updated_count": updated_count,
        "unchanged_ids": [str(timelog_id) for timelog_id in unchanged_ids],
        "failed_ids": [str(timelog_id) for timelog_id in failed_ids],
        "errors": errors,
    }


@router.get("/", response_model=ListResponse)
async def list_timelogs(
    session: SessionDep,
    page: int = Query(1, ge=1),
    size: int = Query(50, ge=1, le=500),
    start_date: date | None = None,
    end_date: date | None = None,
    window_start: datetime | None = None,
    window_end: datetime | None = None,
    query: str | None = None,
    tracking_method: str | None = None,
    area_id: UUID | None = None,
    area_name: str | None = None,
    without_area: bool = False,
    task_id: UUID | None = None,
    without_task: bool = False,
    with_task: bool = False,
) -> ListResponse:
    """List timelogs for the local Web UI."""
    if (start_date is None) != (end_date is None):
        raise HTTPException(
            status_code=400,
            detail="start_date and end_date must be provided together.",
        )
    if (start_date is not None or end_date is not None) and (
        window_start is not None or window_end is not None
    ):
        raise HTTPException(
            status_code=400,
            detail="Use either start_date/end_date or window_start/window_end, not both.",
        )
    if without_task and task_id is not None:
        raise HTTPException(
            status_code=400,
            detail="Use either task_id or without_task, not both.",
        )
    if with_task and (without_task or task_id is not None):
        raise HTTPException(
            status_code=400,
            detail="Use only one of task_id, without_task, or with_task.",
        )

    normalized_window_start = to_storage_timezone(window_start) if window_start else None
    normalized_window_end = to_storage_timezone(window_end) if window_end else None
    filters = TimelogQueryFilters(
        start_date=start_date,
        end_date=end_date,
        query=query,
        tracking_method=tracking_method,
        area_id=area_id,
        area_name=area_name,
        without_area=without_area,
        task_id=task_id,
        without_task=without_task,
        with_task=with_task,
        window_start=normalized_window_start,
        window_end=normalized_window_end,
    )
    total_count = await timelog_services.count_timelogs(session, filters=filters)
    rows = await timelog_services.list_timelogs(
        session,
        query=TimelogListInput(
            filters=filters,
            limit=size,
            offset=(page - 1) * size,
        ),
    )
    items = [_timelog_payload(row) for row in rows]
    return ListResponse(
        items=items,
        pagination=Pagination(
            page=page,
            size=size,
            total=total_count,
            pages=math.ceil(total_count / size) if size else 0,
        ),
        meta={
            "start_date": start_date.isoformat() if start_date else None,
            "end_date": end_date.isoformat() if end_date else None,
            "window_start": normalized_window_start.isoformat()
            if normalized_window_start
            else None,
            "window_end": normalized_window_end.isoformat() if normalized_window_end else None,
            "query": query,
            "tracking_method": tracking_method,
            "area_id": str(area_id) if area_id else None,
            "area_name": area_name,
            "without_area": without_area,
            "task_id": str(task_id) if task_id else None,
            "without_task": without_task,
            "with_task": with_task,
            "limit": size,
            "returned_count": len(items),
            "total_count": total_count,
            "truncated": len(items) < total_count,
        },
    )


@router.post("/")
async def create_timelog(
    payload: TimelogCreate,
    session: SessionDep,
) -> dict[str, object]:
    """Create a timelog."""
    try:
        timelog = await timelog_services.create_timelog(
            session,
            payload=TimelogCreateInput(
                title=payload.title,
                start_time=payload.start_time,
                end_time=payload.end_time,
                tracking_method=payload.tracking_method,
                location=payload.location,
                energy_level=payload.energy_level,
                notes=payload.notes,
                area_id=payload.area_id,
                task_id=payload.task_id,
                person_ids=payload.person_ids,
            ),
        )
    except (LookupError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _timelog_payload(timelog)


@router.post("/batch-update")
async def batch_update_timelogs(
    payload: TimelogBatchUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Batch-update timelogs through LifeOS native batch semantics."""
    title: str | None = None
    find_title_text: str | None = None
    replace_title_text = ""
    changes = TimelogUpdateInput()

    if payload.update_type == "title":
        if payload.title is None:
            raise HTTPException(status_code=400, detail="Title update payload is required")
        if payload.title.mode == "find_replace":
            find_title_text = payload.title.find
            replace_title_text = payload.title.value
        else:
            title = payload.title.value
    elif payload.update_type == "task":
        if payload.task is None:
            raise HTTPException(status_code=400, detail="Task update payload is required")
        changes = TimelogUpdateInput(
            task_id=payload.task.task_id,
            clear_task=payload.task.mode == "clear" or payload.task.task_id is None,
        )
    elif payload.update_type == "area":
        if payload.area is None:
            raise HTTPException(status_code=400, detail="Area update payload is required")
        changes = TimelogUpdateInput(
            area_id=payload.area.area_id,
            clear_area=payload.area.area_id is None,
        )
    elif payload.update_type == "people":
        if payload.people is None:
            raise HTTPException(status_code=400, detail="Person update payload is required")
        if payload.people.mode == "add":
            return await _batch_add_timelog_people(
                session,
                timelog_ids=payload.timelog_ids,
                person_ids=payload.people.person_ids,
            )
        changes = TimelogUpdateInput(
            person_ids=payload.people.person_ids,
            clear_people=payload.people.mode == "clear",
        )
    else:
        raise HTTPException(
            status_code=400, detail=f"Unsupported update type: {payload.update_type}"
        )

    try:
        result = await timelog_services.batch_update_timelogs(
            session,
            timelog_ids=payload.timelog_ids,
            changes=TimelogBatchUpdateInput(
                title=title,
                find_title_text=find_title_text,
                replace_title_text=replace_title_text,
                changes=changes,
            ),
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {
        "updated_count": result.updated_count,
        "unchanged_ids": [str(timelog_id) for timelog_id in result.unchanged_ids],
        "failed_ids": [str(timelog_id) for timelog_id in result.failed_ids],
        "errors": list(result.errors),
    }


@router.patch("/{timelog_id}")
async def update_timelog(
    timelog_id: UUID,
    payload: TimelogUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Update a timelog."""
    fields = payload.model_fields_set
    try:
        timelog = await timelog_services.update_timelog(
            session,
            timelog_id=timelog_id,
            changes=TimelogUpdateInput(
                title=payload.title,
                start_time=payload.start_time,
                end_time=payload.end_time,
                tracking_method=payload.tracking_method,
                location=payload.location,
                clear_location="location" in fields and payload.location is None,
                energy_level=payload.energy_level,
                clear_energy_level="energy_level" in fields and payload.energy_level is None,
                notes=payload.notes,
                clear_notes="notes" in fields and payload.notes is None,
                area_id=payload.area_id,
                clear_area="area_id" in fields and payload.area_id is None,
                task_id=payload.task_id,
                clear_task="task_id" in fields and payload.task_id is None,
                person_ids=payload.person_ids,
                clear_people="person_ids" in payload.model_fields_set and payload.person_ids == [],
            ),
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _timelog_payload(timelog)


@router.delete("/{timelog_id}", status_code=204)
async def delete_timelog(
    timelog_id: UUID,
    session: SessionDep,
) -> None:
    """Soft-delete a timelog."""
    try:
        await timelog_services.delete_timelog(session, timelog_id=timelog_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
